### PR TITLE
Workflow Filtering

### PR DESCRIPTION
**Pull request by rafaelraposospot**
# Hey, I just made a Pull Request!

## Checklist for PR author(s)

- [x] Changes are covered by unit test
- [ ] All tests pass
- [x] Code coverage check passes
- [ ] Error handling is tested
- [ ] Errors are handled at the appropriate layer
- [ ] Errors that cannot be handled where they occur are propagated
- [ ] (optional) Changes are covered by system test
- [ ] Relevant documentation updated
- [x] This PR has NO breaking change to public API
- [ ] This PR has breaking change to public API and it is documented

## Checklist for PR reviewer(s)

- [ ] This PR has been incorporated in release note for the coming version
- [ ] Risky changes introduced by this PR have been all considered

### CODE REVIEW

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java [hunk 1]
```diff
@@ -318,7 +338,8 @@ private Response<List<WorkflowInstanceExecutionData>> instances(
         data = storage.executionData(workflowId, start, stop);
       }
     } catch (IOException e) {
-      throw new RuntimeException("Failed to get execution data of workflow " + workflowId.toKey(), e);
+      throw new RuntimeException("Failed to get execution data of workflow " + workflowId.toKey(),
+          e);
```
Comment (narape) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java:342]:
```suggestion
      throw new UncheckedIOException("Failed to get execution data of workflow " + workflowId.toKey(),
          e);
```
There is already an exception created in java to wrap an `IOException` into a Runtime one: https://docs.oracle.com/javase/8/docs/api/java/io/UncheckedIOException.html

Please change this and the others

**Review by narape — COMMENTED** [review]
Some preliminary feedback

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java (hunk 1 above)
Comment (rafaelraposospot) [reply styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java:342]: To be done in another PR since it's unrelated to the Workflow filtering.

File: styx-api-service/src/main/java/com/spotify/styx/api/util/WorkflowFiltering.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+      Collection<Workflow> workflows, Map<String, String> paramFilters){
+    List<Predicate> workflowFilters = createWorkflowFilters(paramFilters);
+
+    return workflows.stream().filter(w -> workflowFilters.stream().allMatch(pre-> pre.test(w))).collect(toList());
+  }
+
+  private static List<Predicate> createWorkflowFilters(Map<String, String> paramFilters){
+    List<Predicate> predicates = new ArrayList<>();
+
+    String deploymentType = paramFilters.get("deployment_type");
```
Comment (lupasarin) [inline styx-api-service/src/main/java/com/spotify/styx/api/util/WorkflowFiltering.java:50]: Here in this class I would use the constant values from an enum that Nelson suggested.

File: styx-api-service/src/main/java/com/spotify/styx/api/util/WorkflowFiltering.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        Instant deploymentTimeBeforeInstant = Instant.parse(deploymentTimeBefore);
+        return deploymentTime.isBefore(deploymentTimeBeforeInstant);
+      };
+      predicates.add(workflowDeploymentTimeBefore);
+    }
+
+    String deploymentTimeAfter = paramFilters.get("deployment_time_after");
+
+    if(!Strings.isNullOrEmpty(deploymentTimeAfter)){
+      Predicate<Workflow> workflowDeploymentTimeBefore = workflow -> {
```
Comment (lupasarin) [inline styx-api-service/src/main/java/com/spotify/styx/api/util/WorkflowFiltering.java:83]: I understand this should be `workflowDeploymentTimeAfter` instead of Before

File: styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    filters.put("deployment_type", "");
+    filters.put("deployment_time_before",TEST_DEPLOYMENT_TIME_BEFORE.toString() );
+    filters.put("deployment_time_after-type", TEST_DEPLOYMENT_TIME_AFTER.toString());
+
+    List<Workflow> workflows = filterWorkflows(workflowCollection, filters);
+
+    assertThat(workflows, equalTo(new ArrayList<>(workflowCollection)));
+  }
+
+}
```
Comment (lupasarin) [inline styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java:172]: Missing EOF

File: styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    workflowCollection.add(Workflow.create(
+        "id-1", FLYTE_WORKFLOW_CONFIGURATION_WITH_DEPLOYMENT_SOURCE
+    ));
+
+    workflowCollection.add(Workflow.create(
+        "id-2", FLYTE_WORKFLOW_CONFIGURATION_WITH_DEPLOYMENT_SOURCE
+    ));
+
+    Map<String, String> emptyFilters = new HashMap<>();
+    emptyFilters.put("deployment_type", "");
```
Comment (lupasarin) [inline styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java:57]: Same regarding using constant values and +1 for using Map.of. Maybe even a handy method to return an empty map in this case and below?

File: styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java [hunk 6]
```diff
@@ -816,14 +829,29 @@ public void shouldReturnWorkflows() throws Exception {
     assertJson(response, "[*]", hasSize(2));
   }
 
+  @Test
+  public void shouldReturnFilteredDeploymentTypeWorkflow() throws Exception {
+    sinceVersion(Api.Version.V3);
+
+    when(storage.workflows()).thenReturn(FLYTE_DEPLOYMENT_WORKFLOWS);
```
Comment (narape) [inline styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java:836]:
```suggestion
    when(storage.workflows()).thenReturn(buildWorkflowsMap(
      createWorkflowWithDeploymentType("test1", "remote-workflow"),
      createWorkflowWithDeploymentType("test2", "remote-workflow"),
      createWorkflowWithDeploymentType("test3", "")
    ));
```
Having this constant up there make it hard to really grasp what is happening and more importantly if this is correct. Here I am suggesting that you create:
* `createWorkflowWithDeploymentType` for creating a workflow with specified deployment type
* `buildWorkflowsMap` this is a small method that takes an array of workflows and return the map as required
```java
private Map<WorkflowId, Workflow> buildWorkflowsMap(Workflow... workflows) {
   return Arrays.stream(workflows).collect(Collectors.toMap(w -> w.id(), w -> w);
}
```
The `...` are in java called [var args](https://www.baeldung.com/java-varargs)

Then when you do the assertion, you can check that only test1 and test2 are in the response but test3 is not

File: styx-api-service/src/main/java/com/spotify/styx/api/util/WorkflowFiltering.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  private static List<Predicate> createWorkflowFilters(Map<FilterParams, String> paramFilters){
+    List<Predicate> predicates = new ArrayList<>();
+
+    String deploymentType = paramFilters.get(DEPLOYMENT_TYPE);
+
+    if(!Strings.isNullOrEmpty(deploymentType)) {
+      Predicate<Workflow> isWorkflowDeploymentType = workflow -> {
+        WorkflowConfiguration workflowConfiguration = workflow.configuration();
```
Comment (lupasarin) [inline styx-api-service/src/main/java/com/spotify/styx/api/util/WorkflowFiltering.java:57]: Now I'm wondering, don't you need to check if `workflowConfiguration` is empty too? And not only `deploymentSource`?

File: styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java (hunk 6 above)
Comment (lupasarin) [reply styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java:836]: I think the idea is that some workflows have deployment type and some deployment time, right? Just leaving a comment here so we keep in mind that we would need to incorporate `deploymentTime` as an extra parameter if we go this route.

File: styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+
+    workflowCollection.add(Workflow.create(
+        "id-2", FLYTE_WORKFLOW_CONFIGURATION_WITH_DEPLOYMENT_SOURCE
+    ));
+
+    Map<FilterParams, String> emptyFilters = Map.of(DEPLOYMENT_TYPE, "",
+        DEPLOYMENT_TIME_BEFORE, "",
+        DEPLOYMENT_TIME_AFTER, "");
+
+    List<Workflow> workflows = filterWorkflows(workflowCollection, emptyFilters);
```
Comment (lupasarin) [inline styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java:65]: [nit] I would call this `result` to make it clearer how `workflows` and `workflowCollection` are different (same in other test cases)

File: styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+
+    assertThat(workflows, equalTo(new ArrayList<>(workflowCollection)));
```
Comment (lupasarin) [inline styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java:67]: Can we avoid recreating the ArrayList?

File: styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+
+    List<Workflow> workflows = filterWorkflows(workflowCollection, emptyFilters);
+
+    assertThat(workflows, empty());
+  }
+
+  @Test
+  public void shouldReturnWorkflowsWithDeploymentType() {
+    Collection<Workflow> workflowCollection = new ArrayList<>();
+    workflowCollection.add(Workflow.create(
```
Comment (lupasarin) [inline styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java:93]: I would add several ones to be able to check which ones are returned and which ones are not. At least one more without deployment_type

File: styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+
+    List<Workflow> workflows = filterWorkflows(workflowCollection, filters);
+
+    assertThat(workflows, equalTo(new ArrayList<>(workflowCollection)));
+  }
+
+  @Test
+  public void shouldReturnWorkflowsWithDeploymentTimeBefore() {
+    Collection<Workflow> workflowCollection = new ArrayList<>();
+    workflowCollection.add(Workflow.create(
```
Comment (lupasarin) [inline styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java:109]: Same regarding having at least one more workflow in the collection with deployment time NOT before

File: styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+
+    List<Workflow> workflows = filterWorkflows(workflowCollection, filters);
+
+    assertThat(workflows, equalTo(new ArrayList<>(workflowCollection)));
+  }
+
+  @Test
+  public void shouldReturnWorkflowsWithDeploymentTimeAfter() {
+    Collection<Workflow> workflowCollection = new ArrayList<>();
+    workflowCollection.add(Workflow.create(
```
Comment (lupasarin) [inline styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java:125]: Same regarding having at least one more workflow in the collection with deployment time NOT after

File: styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+
+    List<Workflow> workflows = filterWorkflows(workflowCollection, filters);
+
+    assertThat(workflows, equalTo(new ArrayList<>(workflowCollection)));
+  }
+
+  @Test
+  public void shouldReturnWorkflowsWithDeploymentTypeDeploymentTimeBeforeAndAfter() {
+    Collection<Workflow> workflowCollection = new ArrayList<>();
+    workflowCollection.add(Workflow.create(
```
Comment (lupasarin) [inline styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java:141]: Same regarding having at least one more workflow in the collection with deployment time NOT before or after

File: styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+
+    List<Workflow> workflows = filterWorkflows(workflowCollection, filters);
+
+    assertThat(workflows, equalTo(new ArrayList<>(workflowCollection)));
+  }
+
+  @Test
+  public void shouldNotReturnWorkflowWithFilterDeploymentType() {
+    Collection<Workflow> workflowCollection = new ArrayList<>();
+    workflowCollection.add(Workflow.create(
```
Comment (lupasarin) [inline styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java:157]: Same regarding having at least one more workflow in the collection without deploymentType

File: styx-api-service/src/main/java/com/spotify/styx/api/util/WorkflowFiltering.java (hunk 7 above)
Comment (rafaelraposospot) [reply styx-api-service/src/main/java/com/spotify/styx/api/util/WorkflowFiltering.java:57]: So WorkflowConfiguration is not an Optional like deploymentSource so we should be safe.

File: styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java (hunk 8 above)
Comment (rafaelraposospot) [reply styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java:65]: Thank you!

File: styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java (hunk 9 above)
Comment (rafaelraposospot) [reply styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java:67]: We can't since the Collection<Workflow> is inside a List<Workflow>

Comment (rafaelraposospot) [reply styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java:67]: Scratch that. We can.

File: styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java (hunk 10 above)
Comment (rafaelraposospot) [reply styx-api-service/src/test/java/com/spotify/styx/api/util/WorkflowFilteringTest.java]: Done.

File: styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java (hunk 6 above)
Comment (rafaelraposospot) [reply styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java]: I've created a helper class to deal with those. Let me know if that works.

File: styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java [hunk 15]
```diff
@@ -833,17 +837,161 @@ public void shouldReturnWorkflows() throws Exception {
...
 
+  @Test
+  public void shouldReturnFilteredDeploymentTimeBeforeWorkflow() throws Exception {
+    sinceVersion(Api.Version.V3);
+
+    when(storage.workflows()).thenReturn(
+        buildWorkflowMap(
+            createWorkflowWithType("id1", "remote-foo"),
+            createWorkflowWithTime("id2", Instant.ofEpochSecond(1609539330)),
+            createWorkflowWithTime("id3", Instant.ofEpochSecond(1643840130))
```
[the remark this answers is not in the document]
Comment (narape) [reply styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java:864]: The same suggestion  for other tests

File: styx-api-service/src/main/java/com/spotify/styx/api/util/QueryParams.java [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+
+  private final String string;
+
+  QueryParams(String string) {
+    this.string = string;
+  }
+
+  public String getString() {
+    return string;
+  }
```
Comment (narape) [inline styx-api-service/src/main/java/com/spotify/styx/api/util/QueryParams.java:37]:
```suggestion
  private final String queryName;

  QueryParams(String queryName) {
    this.queryName = queryName;
  }

  public String getQueryName() {
    return queryName;
  }
```
This feedback was resolved with a :+1: but no applied, so I am bringing it back
> Let's figure out a better name. Here are some suggestions:
> * name
> * paramName
> * queryName

Comment (rafaelraposospot) [reply styx-api-service/src/main/java/com/spotify/styx/api/util/QueryParams.java:37]: Missread that one and changed the wrong thing. Done.

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java [hunk 17]
```diff
@@ -208,14 +218,26 @@ private WorkflowConfiguration readFromJsonWithDefaults(ByteString payload)
...
   }
 
-  private Response<Collection<Workflow>> workflows() {
+  private Response<Collection<Workflow>> workflows(Request request) {
     try {
-      return Response.forPayload(storage.workflows().values());
+      Map<QueryParams, String> paramFilters =
+          Map.of(DEPLOYMENT_TYPE, getFilterParams(request, DEPLOYMENT_TYPE),
+              DEPLOYMENT_TIME_BEFORE, getFilterParams(request, DEPLOYMENT_TIME_BEFORE),
+              DEPLOYMENT_TIME_AFTER, getFilterParams(request, DEPLOYMENT_TIME_AFTER));
```
Comment (narape) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java:226]:
```java
var paramFilters = Stream.of(QueryParams.values())
  .map(e -> getFilterParams(request, e).map(param -> Map.entry(e, param))
  .map(Optional::stream)
  .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue);
```
This version is a bit better because if no query params are passed, then the map would be empty instead of containing params pointing to empty strings

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java [hunk 18]
```diff
@@ -208,14 +218,26 @@ private WorkflowConfiguration readFromJsonWithDefaults(ByteString payload)
...
+      return Response.forPayload(filterWorkflows(workflows, paramFilters));
+
     } catch (IOException e) {
       throw new RuntimeException("Failed to get workflows", e);
     }
   }
 
+  private String getFilterParams(Request request, QueryParams filter) {
+    return request.parameter(filter.getQueryName()).orElse("");
+  }
```
Comment (narape) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java:239]:
```suggestion
  private Optional<String> getFilterParams(Request request, QueryParams filter) {
    return request.parameter(filter.getQueryName());
  }
```
To make it work with my previous comment, maybe now is simple enough that it could be inlined, but I will leave that decision to you

File: styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java [hunk 19]
```diff
@@ -158,6 +165,16 @@ public class WorkflowResourceTest extends VersionedApiTest {
...
+  private static final Workflow EXISTING_FLYTE_WORKFLOW_CONFIGURATION_WITH_DEPLOYMENT_TYPE =
+      Workflow.create("styx.TestEndpoint1", FLYTE_WORKFLOW_CONFIGURATION_WITH_DEPLOYMENT_TYPE);
+
+  private static final Workflow EXISTING_FLYTE_WORKFLOW_CONFIGURATION_WITH_DEPLOYMENT_TIME =
+      Workflow.create("styx.TestEndpoint2", FLYTE_WORKFLOW_CONFIGURATION_WITH_DEPLOYMENT_TIME);
+
+  private static final Map<WorkflowId, Workflow> FLYTE_DEPLOYMENT_WORKFLOWS =
+      Map.of(EXISTING_FLYTE_WORKFLOW_CONFIGURATION_WITH_DEPLOYMENT_TYPE.id(), EXISTING_FLYTE_WORKFLOW_CONFIGURATION_WITH_DEPLOYMENT_TYPE,
+          EXISTING_FLYTE_WORKFLOW_CONFIGURATION_WITH_DEPLOYMENT_TIME.id(), EXISTING_FLYTE_WORKFLOW_CONFIGURATION_WITH_DEPLOYMENT_TIME);
+
```
Comment (narape) [inline styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java:177]:
```suggestion
```
Unused

File: styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java [hunk 20]
```diff
@@ -816,14 +833,202 @@ public void shouldReturnWorkflows() throws Exception {
...
   @Test
   public void shouldFailedToReturnWorkflows() throws Exception {
     sinceVersion(Api.Version.V3);
 
     when(storage.workflows()).thenThrow(new IOException());
 
     Response<ByteString> response = awaitResponse(
-        serviceHelper.request("GET", path("")));
+        serviceHelper.request("GET", path("?deployment_type=")));
```
Comment (narape) [inline styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java:1031]:
```suggestion
        serviceHelper.request("GET", path("")));
```
It seems that we can revert this change

File: styx-common/src/main/java/com/spotify/styx/testdata/TestData.java [hunk 21]
```diff
@@ -219,6 +267,8 @@ public final class TestData {
           .commitSha(VALID_SHA)
           .build();
 
+
+
```
Comment (narape) [inline styx-common/src/main/java/com/spotify/styx/testdata/TestData.java:271]:
```suggestion
```
No need of new lines

**Review by lupasarin — APPROVED** [review]

File: styx-common/src/main/java/com/spotify/styx/testdata/TestData.java [hunk 22]
```diff
@@ -193,7 +193,55 @@ public final class TestData {
                   .repository("some-organisation/some-path-to-repositry")
                   .source("kubernetes-cluster/namespace/resourceName")
                   .build())
-          .deploymentTime(Instant.now())
+          .deploymentTime(TEST_DEPLOYMENT_TIME)
+          .build();
+
+  public static final WorkflowConfiguration FLYTE_WORKFLOW_CONFIGURATION_WITH_DEPLOYMENT_TYPE =
```
[the remark this answers is not in the document]
Comment (rafaelraposospot) [reply styx-common/src/main/java/com/spotify/styx/testdata/TestData.java:199]: Used in WorkflowFilteringTest

File: styx-api-service/src/main/java/com/spotify/styx/api/util/QueryParams.java [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+
+public enum QueryParams {
+  DEPLOYMENT_TYPE("deployment_type"),
+  DEPLOYMENT_TIME_BEFORE("deployment_time_before"),
+  DEPLOYMENT_TIME_AFTER("deployment_time_after");
+
+
+  private final String queryName;
+
+  QueryParams(String queryName) {
```
Comment (narape) [inline styx-api-service/src/main/java/com/spotify/styx/api/util/QueryParams.java:31]:
```suggestion
  private QueryParams(String queryName) {
```

File: styx-api-service/src/main/java/com/spotify/styx/api/util/WorkflowFiltering.java [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+import java.util.function.Predicate;
+import joptsimple.internal.Strings;
+
+public final class WorkflowFiltering {
+
+  private WorkflowFiltering(){}
+
+  public static List<Workflow> filterWorkflows(
+      Collection<Workflow> workflows, Map<QueryParams, String> paramFilters){
+    List<Predicate> workflowFilters = createWorkflowFilters(paramFilters);
```
Comment (narape) [inline styx-api-service/src/main/java/com/spotify/styx/api/util/WorkflowFiltering.java:45]:
```suggestion
    if (paramFilters.isEmpty()) {
      // Nothing to filter on
      return workflows;
    }
    List<Predicate> workflowFilters = createWorkflowFilters(paramFilters);
```
Do less work when nothing to filter on

**Review by narape — APPROVED** [review]

**Review by narape — APPROVED** [review]
